lib/plug/debugger/highlighter.ex

```diff
@@ -0,0 +1,312 @@
+  # Number `5` might become `" 5"` or `"  5"`, depending
+  # on the maximum line number
+  defp pad_line_number(nr, length) do
+    String.pad_leading(to_string(nr), length)
+  end
+
+  defp pad_lines(lines) do
+    line_numbers = Enum.map(lines, fn {nr, _, _} -> nr end)
+    length = max_digits(line_numbers)
+    Enum.map(lines, fn {nr, line, highlight} ->
+      {pad_line_number(nr, length), line, highlight}
+    end)
+  end
+
+  # Stolen from `Plug.Debugger`
+  defp escape_string(string) do
+    string |> to_string() |> String.trim_trailing() |> Plug.HTML.html_escape()
+  end
+
+  defp escape_line({nr, line, highlighted}) do
+    {nr, escape_string(line), highlighted}
+  end
+
+  defp escape_lines(lines) do
+    Enum.map(lines, &escape_line/1)
+  end
+
+  defp plain_lines(lines) do
+    lines
+    |> pad_lines()
+    |> escape_lines()
+  end
+
+  # Currently only an elixir lexer is supported
+  defp lexer_for_file(file) do
+    cond do
+      String.ends_with?(file, ".ex") -> ElixirLexer
+      String.ends_with?(file, ".exs") -> ElixirLexer
+      :otherwise -> nil
+    end
+  end
+
+  defp highlight_snippet_with_lexer(lines, lexer) do
+    if Code.ensure_loaded?(lexer) do
+      # Unzip the 3-tuples
+      line_numbers = Enum.map(lines, fn {nr, _, _} -> nr end)
+      text_lines = Enum.map(lines, fn {_, line, _} -> line end)
+      # `highlight` here has nothing to do with syntax highlighting;
+      # It marks the lines we mush highlight becuase that's where the error occured.
+      highlight = Enum.map(lines, fn {_, _, highlight} -> highlight end)
+      # Join the lines we're about to highlight
+      # Highlighting will be more precise that way.
+      # Highlighting will still be unreliable if there is a token that spans too many lines,
+      # for example, a long heredoc.
+      # This is an argument for highlighting the whole file and extracting the interesting lines later.
+      # TODO: should we do this?
+      text = Enum.join(text_lines, "")
+      # Get a list of tokens. Tokens may span more than one line.
+      tokens = lexer.lex(text)
+      # We must split the list of tokens into a list of lists,
+      # splitting a token in several parts if needed.
+      token_lines = split_into_lines(tokens)
+      # Finally, render the lines into HTML.
+      # The HTML is already escaped, which means we don't escape it again in the template
+      html_lines = Enum.map(token_lines, fn line -> token_line_to_html(line) end)
+      debugger_lines = Enum.zip([line_numbers, html_lines, highlight])
+      pad_lines(debugger_lines)
+    else
+      plain_lines(lines)
+    end
+  end
+
+  defp token_line_to_html(tokens) do
+    html =
+      tokens
+      |> Enum.map(fn token -> HTMLFormatter.format_token(token) end)
+      |> Enum.join("")
+
+    html
+  end
+
+  # Splits a list of tokens into lines.
+  # If necessary, will split a token on linebreaks.
+  defp split_into_lines(tokens) do
```

Comment:
This seems like a lot of work. What is the goal behind this code? Can't we highlight the whole file instead and extract the lines? I assume the line extraction would be done by breaking the highlighted content on `<br />` or similar?

Reply:
Would it be too much to change makeup_elixir to explicitly mark newlines as `<span class="w.nl">\n</span>` or something of sorts?

Reply:
> I am afraid we are relying on the tokens which may be private API, no?

The tokens are explicitly public API. That's how anyone can write their own formatters that operate on the result of Makeup lexers.

> In any case, how do you write line breaks in Makeup? You do have to write them down somewhere, right?

I write them as literal linebreaks inside the token. For example:

```elixir
iex> alias Makeup.Lexers.ElixirLexer
Makeup.Lexers.ElixirLexer
iex> alias Makeup.Formatters.HTML.HTMLFormatter
Makeup.Formatters.HTML.HTMLFormatter
iex(5)> heredoc = ~S("""
...(5)> a line
...(5)> another line
...(5)> yet another line
...(5)> """)
"\"\"\"\na line\nanother line\nyet another line\n\"\"\""
iex> ElixirLexer.lex(heredoc)
[
  {:string, %{language: :elixir},
   ["\"\"\"", 10, 97, 32, 108, 105, 110, 101, 10, 97, 110, 111, 116, 104, 101,
    114, 32, 108, 105, 110, 101, 10, 121, 101, 116, 32, 97, 110, 111, 116, 104,
    101, 114, 32, 108, 105, 110, 101, 10, "\"\"\""]}
]
iex> heredoc |> ElixirLexer.lex() |> HTMLFormatter.format_as_binary()
"<pre class=\"highlight\"><code><span class=\"s\">&quot;&quot;&quot;\na line\nanother line\nyet another line\n&quot;&quot;&quot;</span></code></pre>"
iex(21)>
```

This is an example of Makeup lexing a multiline token. The output of the lexer is a token. The content of the token is an IOList for efficiency reasons (that's why it has all of those integers).

When you format the list of tokens, you get a string. As you can see there, the  formatter simply inserts newline characters inside the span tags. Those newlines are rendered as newlines because they are meant to be inside a `<pre>` tag.

Reply:
> Would it be too much to change makeup_elixir to explicitly mark newlines as <span class="w.nl">\n</span> or something of sorts?

Probably not. In the limit, I can move this complex function into makeup itself and:

1. Lex the input as I usually do
2. As a postprocessing step split the input into lines
3. Add the extra newline tokens

Alternatively, I *might* be able to rework the lexer so that it adds newlines directly, but it seems it would add some complexity.

In the end, allowing linebreaks inside the tokens reduces the complexity of the lexer. If you want tokens without line breaks (like we do here), you have to perform the work somewhere (you can't magically move the compexity away).

I understand you don't want this complex piece of logic in Plug (it's completely orthogonal to what Plug is meant to do). I can move this inside Makeup, and maintaining it will become my responsibility and not yours. Is this an acceptable compromise?

I was thinking of moving this function inside Makeup, because it's already the second project where I need it.

Reply:
> I was thinking of moving this function inside Makeup, because it's already the second project where I need it.

Sounds good to me. But can you please explain why tokenizing newlines separately would be an issue? Can't you do a new combinator and express the current whitespace combinator as a combinator of the `repeat(newline OR whitespace)`?

Reply:
> Can't you do a new combinator and express the current whitespace combinator as a combinator of the repeat(newline OR whitespace)

Yes, I can. But there are other places where I'd had to add the newline combinator, namely heredocs and other sigils

Actually, I've gone back to review the Elixir lexer and tokenzing newlines separately doesn't look as hard as I thought it's be. The function we're discussing dates from the time where Makeup was based on ExSpirit and the lexer was slightly different (although the token format is exactly the same; as I said, it's public API).

*I believe* there are about 3 or 4 lines of code I need to change. `nimble_parsec` allows me to work at a fairly high level of abstraction and reuse some basic combinators over and over. Combined with my baseline level of laziness when typing code, this makes repetition is very low throughout the codebase.

It'll break compatibility (and require some updates on the test suite), but visually (i.e. on an HTML browser) the result will be the same, and as Makeup is still < 1.0, some compatibility breaks are acceptable, and having newlines as their own token makes the token stream eaiser to use.

In any case, this function will be gone from Plug. The very fact that we're having this discussion means I haven't isolated into Makeup some functionality users shouldn't have to care about. Thanks for your feeback!